### PR TITLE
Fly 1.0 (ES5)

### LINKED ISSUE #80: separate Fly and CLIEngine
_Fly_ the utility people can download via `npm i fly` would continue to be this repo.

The top priority is to move `fly.js` which exports the Fly class should into its module `fly-core`.

`cli/*.js` could stay or go into its own `fly-cli` too. 

The least priority is moving `fmt.js` to `fly-fmt`.

### LINKED ISSUE #100: No export with require('Fly')
The following snippet throws a `object is not a function`(iojs v2.3.3) or `Fly is not a function`(iojs v3.2.0) error.

``` js
let Fly = require('fly');
const fly = new Fly();
```

---

[_Discussion on Gitter_](https://gitter.im/flyjs/fly?at=55e655fee5564d685efaff5b)

### LINKED ISSUE #109: Tasks descriptions missing from output for cli.list
When adding @desc comments to tasks they are not printed in the output of `cli.list`, although this behavior is expected according to the [documentation](https://github.com/flyjs/fly/tree/cb53bf771cec252a58e48f280a6e573ad051ca34/docs#features).

``` js
export async function copy() {
    /** @desc copies all files from source to distribution */
    return await this.source(paths.source).target(paths.distribution);
}
```

```
fly --list
Available tasks

  "copy"
```

This appears to be a problem with:
- Missing [unit tests](https://github.com/flyjs/fly/blob/master/test/cli.js#L34)
- babel gobbling up the comments when transpiling
- different babel configs for building fly (hardcoded in package.json, flyfile), fly tests (babel-node, hardcoded in package.json) and comprehending fly-files and fly-plugins (require-hook, hardcoded in cli/spawn)

### LINKED ISSUE #120: Babel 6 + .babelrc -> TypeError: Cannot read property 'error' of undefined
After update to `babel@6` I faced following issue: when I place `.babelrc` file in the root — build fails. When I remove it — `fly` runs fine.

Repo to reproduce: https://github.com/alexfedoseev/fly-babel-6-issue

``` js
/* .babelrc */

{
  "presets": ["es2015", "stage-3"]
}
```

``` js
/* flyfile.js */

export default async function() {
  console.log('!');
}
```

Screen:

screen shot 2015-12-14 at 14 18 26

Stack:

```
$(npm bin)/fly
stack:
  - TypeError: Cannot read property 'error' of undefined
  -     at OptionManager.mergeOptions (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/fly-util/node_modules/babel-core/lib/transformation/file/options/option-manager.js 126:28)
  -     at OptionManager.addConfig (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/fly-util/node_modules/babel-core/lib/transformation/file/options/option-manager.js 107:10)
  -     at OptionManager.findConfigs (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/fly-util/node_modules/babel-core/lib/transformation/file/options/option-manager.js 168:35)
  -     at OptionManager.init (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/fly-util/node_modules/babel-core/lib/transformation/file/options/option-manager.js 229:12)
  -     at compile (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/fly-util/node_modules/babel-core/lib/api/register/node.js 117:22)
  -     at normalLoader (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/fly-util/node_modules/babel-core/lib/api/register/node.js 199:14)
  -     at Object.require.extensions.(anonymous function) [as .js] (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/fly-util/node_modules/babel-core/lib/api/register/node.js 216:7)
  -     at Module.load (module.js 356:32)
  -     at Function.Module._load (module.js 313:12)
  -     at Module.require (module.js 366:17)
  -     at require (module.js 385:17)
  -     at Object.spawn$ (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/fly/dist/cli/spawn.js 39:29)
  -     at tryCatch (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/babel-runtime/regenerator/runtime.js 72:40)
  -     at GeneratorFunctionPrototype.invoke [as _invoke] (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/babel-runtime/regenerator/runtime.js 334:22)
  -     at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/babel-runtime/regenerator/runtime.js 105:21)
  -     at onFulfilled (~/Dev/Sandboxes/tmp/fly-babel-6-issue/node_modules/co/index.js 65:19)
message: Cannot read property 'error' of undefined
```

### LINKED ISSUE #140: TypeError: this.source(...).less is not a function
`TypeError: this.source(...).less is not a function` using your example.

### LINKED ISSUE #141: message: Maximum call stack size exceeded
Hi. I try to use Fly v0.9.1, but it doesn't work in part of environments Node v4 and npm v2.
I develop [fly-ava](https://github.com/pine613/fly-ava) plugins, and also use it, so I've found errors as the following.

```
message: Maximum call stack size exceeded
stack: 
  - RangeError: Maximum call stack size exceeded
  -     at Function.isArray (native)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 37:13)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
  -     at reduce (~/build/pine613/fly-ava/node_modules/fly/dist/utils/bind.js 43:14)
```

https://travis-ci.org/pine613/fly-ava/jobs/110563430

I think that `reduce` util function has bugs.
https://github.com/bucaran/fly/blob/master/src/utils/bind.js#L21-L33

This code runs forever.

```
reduce([]);
```

Thank you.

### PR DESCRIPTION
**Pull request by lukeed**
## Changes
- Rewrites everything into ES5, dropping Babel as a core dependency.
- Flyfiles must be written in ES5 (module.exports & generators/yield only). 
- Flyfiles must be named `flyfile.js`. "Flyfile.js" is no longer recognized.
- Removes some old fly-util helpers in favor of NPM packages: `expand`, `flatten`, `filter`
- Added new `util` helpers: `find`, `read`
- Moved `src` directory to `lib` and removed `dist` since there is no longer a build step.
- Moved `bin/index` to `cli.js`
- Changed ESLint styling. Now uses out-of-the-box setup from [XO](https://github.com/sindresorhus/xo)
- Custom timestamp() function for CLI timestamps, in place of heavy `dateformat` dep.
- Dropped `mz` dep. Wrapped `fs.writeFile` and `fs.readFile` with `Promise` instead
- Task descriptions re-enabled. Closes #109
- Export Fly from main CLI file. Closes #100
## Metrics
- CLI executable runs in `~130ms` compared to `1.4s` with Babel
- The slowest test w/ Travis completes in `1m 39s` compared to `4m` with Babel (included compile step)
- More reliable test comparison: Full test suite completes in `~570ms` compared to `1.8s` with Babel (no compile step)
## Babel-related Issues to Close

Closes #141. Closes #120. Closes #141.
## Other Closed Issues

Closes #140. Closes #80.
## Future
- Extract plugin logic to `fly-plugins`
-  Add new plugin 'mounting' structure to fly-plugin for more flexible & easier plugin creation.
- Create `fly-esnext` to allow "native" ES6/7 support (flyfiles + plugins)
#### Notes

`cli.options` only catches the first task. Eg: `npm run fly lint scripts` will only register `scripts`. Issue opened [here](https://github.com/bucaran/parsec/issues/9)

Do we really need interpret? (`utils.bind`)

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,12 @@
+root = true
+
+[*]
+indent_style = tab
```
Comment (ghost) [inline .editorconfig:4]: Nice.

File: .eslintrc [hunk 2]
```diff
@@ -1,20 +0,0 @@
-{
```
Comment (ghost) [inline .eslintrc:1]: So we dropped .eslintrc in favor of?

Comment (lukeed) [inline .eslintrc:1]: XO. It's essentially a predefined set of ESLint rules

Comment (lukeed) [inline .eslintrc:1]: AVA only was installed for a second. I was going to replace Tape with AVA but it wasn't worth it.

File: test/fly.js [hunk 3]
```diff
@@ -1,307 +1,349 @@
...
+		t.ok(true, 'throw an error if filter already exists');
+	}
+
+	fly.source('');
+	t.deepEqual(fly._.filters, [], 'reset filter each time source is called');
+
+	t.end();
+});
+
+// test('✈  fly.watch', function (t) {
```
Comment (ghost) [inline test/fly.js]: I am noticing large commented blanks like this, any actual comments? :smile:

File: lib/utils/defer.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Promisify a function of the form (value, options, cb)
+ * @param  {Function} asyncFunc The function
+ * @return {Function} 					The new function, as a Promise
```
Comment (ghost) [inline lib/utils/defer.js:4]: What's with the funny indentation in the comments?

File: .eslintrc (hunk 2 above)
Comment (ghost) [inline .eslintrc:1]: I meant XO, sorry.

Comment (ghost) [inline .eslintrc:1]: I don't like semicolons, is it possible to make XO not complain about it?

File: lib/utils/defer.js (hunk 4 above)
Comment (lukeed) [inline lib/utils/defer.js:4]: It's the tabbing. JSDocs pushes over the description so that it lines up with the previous line's description

File: .eslintrc (hunk 2 above)
Comment (lukeed) [inline .eslintrc:1]: It is, but no semicolon usage makes for inconsistent reading in the earlier Node versions. You got away with in the previous versions because Babel added them in accordingly

File: test/fly.js (hunk 3 above)
Comment (lukeed) [inline test/fly.js]: Looks like I overlooked this one. I remember the test never finished, which means I forgot a `yield` or a generator function inside the test. 

Fixing it now

File: lib/utils/defer.js (hunk 4 above)
Comment (ghost) [inline lib/utils/defer.js:4]: That's hard to live with :crying_cat_face:

File: .eslintrc (hunk 2 above)
Comment (ghost) [inline .eslintrc:1]: > It is, but no semicolon usage makes for inconsistent reading in the earlier Node versions. 

I was not aware of this. How early? If you mean < 10 I don't mind.

File: lib/utils/defer.js (hunk 4 above)
Comment (lukeed) [inline lib/utils/defer.js:4]: I can fix it :smile: It looks awful in github, but works nicely inside an editor.

screen shot 2016-03-13 at 2 11 10 pm

Comment (ghost) [inline lib/utils/defer.js:4]: Oh, ok, so it's just on GitHub. Then nevermind.

Comment (ghost) [inline lib/utils/defer.js:4]: I followed JSDocs style in the original, but are you using JSDocs now to generate the docs?

File: .eslintrc (hunk 2 above)
Comment (lukeed) [inline .eslintrc:1]: I don't remember... Think I just do it cuz it's customary / "politically correct" javascript lol.

It's tedious but I'll remove 'em and give it a spin in various envs.

File: lib/utils/defer.js (hunk 4 above)
Comment (lukeed) [inline lib/utils/defer.js:4]: Yessir

File: .eslintrc (hunk 2 above)
Comment (ghost) [inline .eslintrc:1]: This is my only pet peeve. I'm going over the code rewrites now.

Comment (lukeed) [inline .eslintrc:1]: That's ok, I'll defer :stuck_out_tongue_closed_eyes: 

As long as the spacing/line breaks stay, I'm happy. It was so hard to follow the function chaining & parentheses wrapping before. Basically relied on my editor to track & highlight the pairs every time.

Comment (ghost) [inline .eslintrc:1]: :+1: 

I like your code a lot, it feels more mature than my original version and it's more similar to how I write these days. I still like to do a hack here and there, but I tend to favor more lines if it helps readability these days more.

**Comment by ghost** [comment]
I'll keep going over the entire code, this will take me another hour or so, then I'll merge.

**Comment by lukeed** [comment]
Okie doke. I'm going to push over the no-semicolon version. Publish a `beta5` and test it locally in projects under all major environments.

**Comment by lukeed** [comment]
It doesn't work man... tests start blowing up as soon as the semicolons (inside `tests/`) are removed.

I mean, it _could_ be just a Tape cli thing... but it seems like a bad omen.

**Comment by lukeed** [comment]
Without semicolons, Node doesn't like the inline logic stuff.

For instance, I had to change 

``` js
(self.events[name] || []).forEach( // ...
```

and pull it out as a variable before running.

``` js
var evts = self.events[name] || []
evts.forEach( //..
```

Now, similarly, I'm noticing that all the tests with similar array prototype functions also fail. Eg

``` js
// fails
t.ok(cli !== undefined, 'is defined')
['help', 'list', 'spawn', 'version'].forEach( //...

// passes
t.ok(cli !== undefined, 'is defined');
['help', 'list', 'spawn', 'version'].forEach( //...

// passes
t.ok(cli !== undefined, 'is defined')
var methods = ['help', 'list', 'spawn', 'version']
methods.forEach( //...
```

**Comment by ghost** [comment]
> Without semicolons, Node doesn't like the inline logic stuff.

Well, yeah, that's the only _no-semicolon_ catch. :lollipop:

**Comment by lukeed** [comment]
Ah :flushed: didnt know this, never not using semicolons :stuck_out_tongue_winking_eye:

**Comment by ghost** [comment]
Well, I can always fix that later. Unless you are sure you got it working, let's just forget about it for now.

**Comment by ghost** [comment]
OK this looks great. Just one more thing, would you mind squashing all those commits into one and add the commit messages as bullet points to a large commit?

**Comment by lukeed** [comment]
:sob: sounds ridiculous, but I was counting on all these for my github activity/history. first time i've really contributed to something. Basically the same "self-promotive" reasoning that has you keeping Fly under your personal account.

**Comment by ghost** [comment]
I just prefer few large commits than having micro commits, but if this means so much to you, no problemo! :smile: 

> Basically the same "self-promotive" reasoning that has you keeping Fly under your personal account.

That's where it started. I then moved to FlyJS org, but TBH since the project didn't get enough contributors I moved it back home. I think it was already here when you came in.

**Comment by lukeed** [comment]
Thank you, sir :bow: Mostly just because it spanned a few days

**Comment by ghost** [comment]
Come to think about it, my request was not at all thoughtful, if you had so many commits you probably meant to have them, so I apologize for even asking you to squash them into a single commit.

File: .eslintrc (hunk 2 above)
Comment (lukeed) [inline .eslintrc:1]: Thanks! I used to write how you did/do, but I found it really problematic when I came back to an old project and had a hard time reading thru what I did

**Comment by lukeed** [comment]
It's ok! I'm all for being succinct too. 

And no, I wasn't planning on having a large number of commits. That's silly. I just followed the practice of consolidating each tangible step into a commit

**Comment by lukeed** [comment]
@bucaran so if you think this is ok to merge, I have about an hour to update the docs & an example or two

**Comment by lukeed** [comment]
:raised_hands:

**Comment by ghost** [comment]
You're awesome @lukeed! Thanks for making Fly 1.0 happen.

Hey dude @watilde, finally we have 1.0 and guess who's the mastermind behind it?

**Comment by watilde** [comment]
Woooo kewl, awesome work! I saw a history. ":tada:" to all of you hackers!

**Comment by ghost** [comment]
Thanks @watilde!
